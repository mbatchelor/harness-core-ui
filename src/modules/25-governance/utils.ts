import type { GovernanceMetadata } from 'services/cd-ng'

interface DoesGovernanceMetaDataHasErrorOrWarningRtn {
  governanceMetaDataHasError: boolean
  governanceMetaDataHasWarning: boolean
}
export const doesGovernanceHasErrorOrWarning = (
  governanceMetadata: GovernanceMetadata | undefined
): DoesGovernanceMetaDataHasErrorOrWarningRtn => {
  let governanceMetaDataHasWarning = false
  let governanceMetaDataHasError = false
  if (governanceMetadata && governanceMetadata?.status === 'error') {
    governanceMetaDataHasError = true
  }
  if (governanceMetadata && governanceMetadata?.status === 'warning') {
    governanceMetaDataHasWarning = true
  }
  return { governanceMetaDataHasError, governanceMetaDataHasWarning }
}
