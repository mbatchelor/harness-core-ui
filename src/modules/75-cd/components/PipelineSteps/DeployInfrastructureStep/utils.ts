/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { FormikProps } from 'formik'
import { isEmpty } from 'lodash-es'

import { getMultiTypeFromValue, MultiTypeInputType } from '@harness/uicore'
import type { EnvironmentResponseDTO, PipelineInfrastructure } from 'services/cd-ng'

import type { StepViewType } from '@pipeline/components/AbstractSteps/Step'

export interface DeployInfrastructureProps {
  initialValues: PipelineInfrastructure
  onUpdate?: (data: PipelineInfrastructure) => void
  readonly: boolean
  allowableTypes: MultiTypeInputType[]
  stepViewType?: StepViewType
  inputSetData?: {
    template?: PipelineInfrastructure
    path?: string
    readonly?: boolean
  }
}

export interface DeployInfrastructureState {
  isEdit: boolean
  isEnvironment: boolean
  formik?: FormikProps<PipelineInfrastructure>
  data?: EnvironmentResponseDTO
}

export function isEditEnvironment(data: PipelineInfrastructure): boolean {
  if (getMultiTypeFromValue(data.environmentRef) !== MultiTypeInputType.RUNTIME && !isEmpty(data.environmentRef)) {
    return true
  } else if (data.environment && !isEmpty(data.environment.identifier)) {
    return true
  }
  return false
}

export function isEditInfrastructure(data: any): boolean {
  if (
    getMultiTypeFromValue(data.infrastructureRef) !== MultiTypeInputType.RUNTIME &&
    !isEmpty(data.infrastructureRef)
  ) {
    return true
  } else if (data.infrastructure && !isEmpty(data.infrastructure.identifier)) {
    return true
  }
  return false
}
