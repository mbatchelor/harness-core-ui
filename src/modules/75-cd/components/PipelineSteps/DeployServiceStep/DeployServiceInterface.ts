/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { MultiTypeInputType } from '@harness/uicore'
import type { FormikProps } from 'formik'
import type { StepViewType } from '@pipeline/components/AbstractSteps/Step'
import type { ServiceConfig, ServiceDefinition, ServiceRequestDTO, ServiceResponseDTO } from 'services/cd-ng'

export interface DeployServiceProps {
  initialValues: DeployServiceData
  onUpdate?: (data: DeployServiceData) => void
  stepViewType?: StepViewType
  readonly: boolean
  inputSetData?: {
    template?: DeployServiceData
    path?: string
    readonly?: boolean
  }
  allowableTypes: MultiTypeInputType[]
  serviceLabel?: string
}

export interface DeployServiceState {
  isEdit: boolean
  data?: ServiceResponseDTO
  isService: boolean
  formik?: FormikProps<DeployServiceData>
}

export interface DeployServiceData extends Omit<ServiceConfig, 'serviceRef'> {
  serviceRef?: string
  isNewServiceEntity?: boolean
  deploymentType?: ServiceDefinition['type']
}

export interface NewEditServiceModalProps {
  isEdit: boolean
  isService: boolean
  data: ServiceResponseDTO
  serviceIdentifier?: string
  onCreateOrUpdate(data: ServiceRequestDTO): void
  closeModal?: () => void
}
