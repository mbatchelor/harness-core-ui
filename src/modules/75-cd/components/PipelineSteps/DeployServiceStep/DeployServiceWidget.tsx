/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { useEffect, useMemo, useState } from 'react'
import {
  ButtonSize,
  ButtonVariation,
  Formik,
  FormInput,
  getMultiTypeFromValue,
  Dialog,
  Layout,
  MultiTypeInputType,
  SelectOption,
  shouldShowError
} from '@harness/uicore'
import { useModalHook } from '@harness/use-modal'
import * as Yup from 'yup'
import { defaultTo, isEmpty, isNil, noop, omit } from 'lodash-es'
import { useParams } from 'react-router-dom'
import type { FormikProps, FormikValues } from 'formik'
import type { IDialogProps } from '@blueprintjs/core'
import produce from 'immer'
import { ServiceRequestDTO, ServiceResponseDTO, ServiceYaml, useGetServiceList, useGetServiceV2 } from 'services/cd-ng'
import { useStrings } from 'framework/strings'
import type { PipelineType } from '@common/interfaces/RouteInterfaces'
import { useToaster } from '@common/exports'
import { useVariablesExpression } from '@pipeline/components/PipelineStudio/PiplineHooks/useVariablesExpression'
import { ResourceType } from '@rbac/interfaces/ResourceType'
import { PermissionIdentifier } from '@rbac/interfaces/PermissionIdentifier'
import useRBACError from '@rbac/utils/useRBACError/useRBACError'
import { StageErrorContext } from '@pipeline/context/StageErrorContext'
import { DeployTabs } from '@pipeline/components/PipelineStudio/CommonUtils/DeployStageSetupShellUtils'
import { getServiceRefSchema } from '@cd/components/PipelineSteps/PipelineStepsUtil'
import RbacButton from '@rbac/components/Button/Button'
import ServiceEntityEditModal from '@cd/components/Services/ServiceEntityEditModal/ServiceEntityEditModal'
import { usePipelineContext } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'
import type { DeployServiceData, DeployServiceProps, DeployServiceState } from './DeployServiceInterface'
import { flexStart, isEditService } from './DeployServiceUtils'
import { NewEditServiceModal } from './NewEditServiceModal'
import css from './DeployServiceStep.module.scss'

function DeployServiceWidget({
  initialValues,
  onUpdate,
  readonly,
  allowableTypes,
  serviceLabel
}: DeployServiceProps): React.ReactElement {
  const { getString } = useStrings()
  const {
    state: {
      pipeline,
      selectionState: { selectedStageId }
    }
  } = usePipelineContext()

  const { accountId, projectIdentifier, orgIdentifier } = useParams<
    PipelineType<{
      orgIdentifier: string
      projectIdentifier: string
      pipelineIdentifier: string
      accountId: string
    }>
  >()
  const queryParams = useMemo(
    () => ({
      accountIdentifier: accountId,
      orgIdentifier,
      projectIdentifier
    }),
    [accountId, orgIdentifier, projectIdentifier]
  )
  const { showError } = useToaster()
  const { getRBACErrorMessage } = useRBACError()
  const { expressions } = useVariablesExpression()
  const serviceRef = initialValues?.service?.identifier || initialValues?.serviceRef

  const [services, setService] = useState<ServiceYaml[]>()
  const [selectOptions, setSelectOptions] = useState<SelectOption[]>()
  const [state, setState] = useState<DeployServiceState>({ isEdit: false, isService: false })
  const [type, setType] = useState<MultiTypeInputType>(getMultiTypeFromValue(serviceRef))

  const { subscribeForm, unSubscribeForm } = React.useContext(StageErrorContext)
  const formikRef = React.useRef<FormikProps<unknown> | null>(null)
  const isNewServiceEntity = (): boolean => {
    return !!initialValues.isNewServiceEntity
  }
  const {
    data: serviceResponse,
    error,
    loading
  } = useGetServiceList({
    queryParams: {
      ...queryParams,
      type: isNewServiceEntity() ? initialValues.deploymentType : undefined
    }
  })

  const {
    data: selectedServiceResponse,
    refetch: refetchServiceData,
    loading: serviceDataLoading
  } = useGetServiceV2({
    serviceIdentifier: '',
    queryParams,
    lazy: true
  })

  useEffect(() => {
    if (!isNil(selectOptions) && initialValues.serviceRef) {
      if (getMultiTypeFromValue(initialValues.serviceRef) === MultiTypeInputType.FIXED) {
        const doesExist = selectOptions.filter(service => service.value === initialValues.serviceRef).length > 0
        if (!doesExist) {
          if (!readonly) {
            formikRef.current?.setFieldValue('serviceRef', '')
          } else {
            const options = [...selectOptions]
            options.push({
              label: initialValues.serviceRef,
              value: initialValues.serviceRef
            })
            setSelectOptions(options)
          }
        }
      }
    }
  }, [selectOptions])

  useEffect(() => {
    if (!isNil(services)) {
      setSelectOptions(
        services.map(service => {
          return { label: service.name, value: service.identifier }
        })
      )
    }
  }, [services])

  useEffect(() => {
    if (!loading) {
      let serviceList: ServiceYaml[] = []
      if (serviceResponse?.data?.content?.length) {
        serviceList = serviceResponse.data.content.map(service => ({
          identifier: defaultTo(service.service?.identifier, ''),
          name: defaultTo(service.service?.name, ''),
          description: service.service?.description,
          tags: service.service?.tags
        }))
      }
      if (initialValues.service) {
        const { identifier } = initialValues.service
        const isExist = serviceList.some(service => service.identifier === identifier)
        if (identifier && !isExist) {
          serviceList.push({
            identifier: defaultTo(initialValues.service?.identifier, ''),
            name: defaultTo(initialValues.service?.name, ''),
            description: initialValues.service?.description,
            tags: initialValues.service?.tags
          })
        }
      }
      setService(serviceList)
    }
  }, [loading, serviceResponse?.data?.content?.length])

  useEffect(() => {
    subscribeForm({ tab: DeployTabs.SERVICE, form: formikRef })
    return () => unSubscribeForm({ tab: DeployTabs.SERVICE, form: formikRef })
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  if (error?.message) {
    if (shouldShowError(error)) {
      showError(getRBACErrorMessage(error))
    }
  }

  const updateServicesList = (value: ServiceRequestDTO): void => {
    formikRef.current?.setValues({ serviceRef: value.identifier, ...(state.isService && { service: {} }) })
    if (!isNil(services)) {
      const newService = {
        identifier: defaultTo(value.identifier, ''),
        name: defaultTo(value.name, ''),
        description: value.description,
        tags: value.tags
      }
      const newServicesList = [...services]
      const existingIndex = newServicesList.findIndex(item => item.identifier === value.identifier)
      if (existingIndex >= 0) {
        newServicesList.splice(existingIndex, 1, newService)
      } else {
        newServicesList.unshift(newService)
      }
      setService(newServicesList)
    }
  }

  const onServiceEntityCreate = (newServiceInfo: ServiceYaml): void => {
    hideModal()
    formikRef.current?.setValues({ serviceRef: newServiceInfo.identifier, ...(state.isService && { service: {} }) })
    const newServiceData = produce(services, draft => {
      draft?.unshift({
        identifier: newServiceInfo.identifier,
        name: newServiceInfo.name
      })
    })
    setService(newServiceData)
  }

  const onServiceChange = (
    fieldValue: SelectOption,
    formikValue: any,
    setFormikValue: (field: string, value: any) => void
  ): void => {
    if (formikValue.service?.identifier && fieldValue.value !== formikValue.service.identifier) {
      setService(services?.filter(service => service.identifier !== formikValue.service?.identifier))
      setFormikValue('service', undefined)
    }
  }

  const editService = (values: FormikValues): void => {
    if (initialValues.isNewServiceEntity) {
      refetchServiceData({
        pathParams: {
          serviceIdentifier: values.serviceRef
        },
        queryParams
      })
    }
    if (values.service?.identifier) {
      setState({
        isEdit: true,
        isService: true,
        data: values.service
      })
    } else {
      setState({
        isEdit: true,
        isService: false,
        data: services?.find(service => service.identifier === values.serviceRef)
      })
    }
    showModal()
  }

  const DIALOG_PROPS: IDialogProps = {
    isOpen: true,
    usePortal: true,
    autoFocus: true,
    canEscapeKeyClose: false,
    canOutsideClickClose: false,
    enforceFocus: false,
    className: isNewServiceEntity() ? css.editServiceDialog : '',
    style: isNewServiceEntity() ? { width: 1114 } : {}
  }
  const serviceEntityProps = state.isEdit
    ? {
        serviceResponse: selectedServiceResponse?.data?.service as ServiceResponseDTO,
        isLoading: serviceDataLoading,
        serviceCacheKey: `${pipeline.identifier}-${selectedStageId}-service`
      }
    : {}
  const [showModal, hideModal] = useModalHook(
    () => (
      <Dialog
        onClose={onClose}
        title={state.isEdit ? getString('editService') : getString('newService')}
        {...DIALOG_PROPS}
      >
        {isNewServiceEntity() ? (
          <ServiceEntityEditModal
            {...serviceEntityProps}
            onCloseModal={hideModal}
            onServiceCreate={onServiceEntityCreate}
            isServiceCreateModalView={!state.isEdit}
          />
        ) : (
          <NewEditServiceModal
            data={{
              name: defaultTo(state.data?.name, ''),
              identifier: defaultTo(state.data?.identifier, ''),
              orgIdentifier,
              projectIdentifier,
              ...state.data
            }}
            isEdit={state.isEdit}
            isService={state.isService}
            onCreateOrUpdate={value => {
              updateServicesList(value)
              onClose.call(null)
            }}
            closeModal={onClose}
          />
        )}
      </Dialog>
    ),
    [state, selectedServiceResponse]
  )

  const onClose = React.useCallback(() => {
    setState({ isEdit: false, isService: false })
    hideModal()
  }, [hideModal])

  return (
    <>
      <Formik<DeployServiceData>
        formName="deployServiceStepForm"
        onSubmit={noop}
        validate={values => {
          if (!isEmpty(values.service)) {
            onUpdate?.({ ...omit(values, 'serviceRef') })
          } else {
            onUpdate?.({ ...omit(values, 'service'), serviceRef: values.serviceRef })
          }
        }}
        initialValues={{
          ...initialValues,
          ...{ serviceRef }
        }}
        validationSchema={Yup.object().shape({
          serviceRef: getServiceRefSchema(getString)
        })}
      >
        {formik => {
          window.dispatchEvent(new CustomEvent('UPDATE_ERRORS_STRIP', { detail: DeployTabs.SERVICE }))
          formikRef.current = formik as FormikProps<unknown> | null
          const { values, setFieldValue } = formik
          return (
            <Layout.Horizontal
              className={css.formRow}
              spacing="medium"
              flex={{ alignItems: flexStart, justifyContent: flexStart }}
            >
              <FormInput.MultiTypeInput
                tooltipProps={{ dataTooltipId: 'specifyYourService' }}
                label={serviceLabel ? serviceLabel : getString('cd.pipelineSteps.serviceTab.specifyYourService')}
                name="serviceRef"
                useValue
                disabled={readonly || (type === MultiTypeInputType.FIXED && loading)}
                placeholder={loading ? getString('loading') : getString('cd.pipelineSteps.serviceTab.selectService')}
                multiTypeInputProps={{
                  onTypeChange: setType,
                  width: 300,
                  expressions,
                  onChange: val => onServiceChange(val as SelectOption, values, setFieldValue),
                  selectProps: {
                    disabled: loading,
                    addClearBtn: true && !readonly,
                    items: selectOptions || []
                  },
                  allowableTypes
                }}
                selectItems={selectOptions || []}
              />
              {isEditService(values) && !loading ? (
                <RbacButton
                  size={ButtonSize.SMALL}
                  text={getString('editService')}
                  variation={ButtonVariation.LINK}
                  id="edit-service"
                  disabled={readonly}
                  permission={{
                    permission: PermissionIdentifier.EDIT_SERVICE,
                    resource: {
                      resourceType: ResourceType.SERVICE,
                      resourceIdentifier: services ? (services[0]?.identifier as string) : ''
                    },
                    options: {
                      skipCondition: ({ resourceIdentifier }) => !resourceIdentifier
                    }
                  }}
                  onClick={() => editService(values)}
                />
              ) : (
                <RbacButton
                  size={ButtonSize.SMALL}
                  text={getString('cd.pipelineSteps.serviceTab.plusNewService')}
                  variation={ButtonVariation.LINK}
                  id="add-new-service"
                  disabled={readonly}
                  permission={{
                    permission: PermissionIdentifier.EDIT_SERVICE,
                    resource: {
                      resourceType: ResourceType.SERVICE
                    }
                  }}
                  onClick={() => {
                    setState({
                      isEdit: false,
                      isService: false
                    })
                    showModal()
                  }}
                />
              )}
            </Layout.Horizontal>
          )
        }}
      </Formik>
    </>
  )
}
export default DeployServiceWidget
