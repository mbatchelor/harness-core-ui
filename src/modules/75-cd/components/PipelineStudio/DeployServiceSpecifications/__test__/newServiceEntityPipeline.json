{
  "state": {
    "pipeline": {
      "name": "serviceentity-pipeline",
      "identifier": "serviceentitypipeline",
      "projectIdentifier": "Bhavya",
      "orgIdentifier": "default",
      "tags": {},
      "stages": [
        {
          "stage": {
            "name": "deploy",
            "identifier": "deploy",
            "description": "",
            "type": "Deployment",
            "spec": {
              "deploymentType": "Kubernetes",
              "service": {
                "serviceRef": "servicebhavya",
                "serviceInputs": {
                  "serviceDefinition": {
                    "type": "Kubernetes",
                    "spec": {
                      "manifests": [
                        {
                          "manifest": {
                            "identifier": "manifest1",
                            "type": "K8sManifest",
                            "spec": {
                              "store": {
                                "type": "Github",
                                "spec": {
                                  "connectorRef": "<+input>",
                                  "repoName": "<+input>"
                                }
                              }
                            }
                          }
                        }
                      ],
                      "artifacts": {
                        "primary": {
                          "type": "Gcr",
                          "spec": {
                            "connectorRef": "<+input>",
                            "imagePath": "<+input>",
                            "tag": "<+input>"
                          }
                        }
                      }
                    }
                  }
                }
              },
              "environment": {
                "environmentRef": "<+input>",
                "deployToAll": false,
                "environmentInputs": "<+input>",
                "serviceOverrideInputs": "<+input>",
                "infrastructureDefinitions": "<+input>"
              },
              "execution": {
                "steps": [
                  {
                    "step": {
                      "name": "Rollout Deployment",
                      "identifier": "rolloutDeployment",
                      "type": "K8sRollingDeploy",
                      "timeout": "10m",
                      "spec": {
                        "skipDryRun": false
                      }
                    }
                  }
                ],
                "rollbackSteps": [
                  {
                    "step": {
                      "name": "Rollback Rollout Deployment",
                      "identifier": "rollbackRolloutDeployment",
                      "type": "K8sRollingRollback",
                      "timeout": "10m",
                      "spec": {}
                    }
                  }
                ]
              }
            },
            "tags": {},
            "failureStrategies": [
              {
                "onFailure": {
                  "errors": ["AllErrors"],
                  "action": {
                    "type": "StageRollback"
                  }
                }
              }
            ]
          }
        }
      ]
    },
    "originalPipeline": {
      "name": "serviceentity-pipeline",
      "identifier": "serviceentitypipeline",
      "projectIdentifier": "Bhavya",
      "orgIdentifier": "default",
      "tags": {},
      "stages": [
        {
          "stage": {
            "name": "deploy",
            "identifier": "deploy",
            "description": "",
            "type": "Deployment",
            "spec": {
              "deploymentType": "Kubernetes",
              "service": {
                "serviceRef": "servicebhavya",
                "serviceInputs": {
                  "serviceDefinition": {
                    "type": "Kubernetes",
                    "spec": {
                      "manifests": [
                        {
                          "manifest": {
                            "identifier": "manifest1",
                            "type": "K8sManifest",
                            "spec": {
                              "store": {
                                "type": "Github",
                                "spec": {
                                  "connectorRef": "<+input>",
                                  "repoName": "<+input>"
                                }
                              }
                            }
                          }
                        }
                      ],
                      "artifacts": {
                        "primary": {
                          "type": "Gcr",
                          "spec": {
                            "connectorRef": "<+input>",
                            "imagePath": "<+input>",
                            "tag": "<+input>"
                          }
                        }
                      }
                    }
                  }
                }
              },
              "environment": {
                "environmentRef": "<+input>",
                "deployToAll": false,
                "environmentInputs": "<+input>",
                "serviceOverrideInputs": "<+input>",
                "infrastructureDefinitions": "<+input>"
              },
              "execution": {
                "steps": [
                  {
                    "step": {
                      "name": "Rollout Deployment",
                      "identifier": "rolloutDeployment",
                      "type": "K8sRollingDeploy",
                      "timeout": "10m",
                      "spec": {
                        "skipDryRun": false
                      }
                    }
                  }
                ],
                "rollbackSteps": [
                  {
                    "step": {
                      "name": "Rollback Rollout Deployment",
                      "identifier": "rollbackRolloutDeployment",
                      "type": "K8sRollingRollback",
                      "timeout": "10m",
                      "spec": {}
                    }
                  }
                ]
              }
            },
            "tags": {},
            "failureStrategies": [
              {
                "onFailure": {
                  "errors": ["AllErrors"],
                  "action": {
                    "type": "StageRollback"
                  }
                }
              }
            ]
          }
        }
      ]
    },
    "pipelineIdentifier": "serviceentitypipeline",
    "pipelineView": {
      "isSplitViewOpen": true,
      "isDrawerOpened": false,
      "isYamlEditable": false,
      "splitViewData": {
        "type": "StageView"
      },
      "drawerData": {
        "type": "AddCommand"
      }
    },
    "schemaErrors": false,
    "storeMetadata": {},
    "gitDetails": {},
    "entityValidityDetails": {
      "valid": true,
      "invalidYaml": null
    },
    "templateTypes": {},
    "isLoading": false,
    "isBEPipelineUpdated": false,
    "isDBInitialized": true,
    "isUpdated": false,
    "isInitialized": true,
    "selectionState": {
      "selectedStageId": "deploy",
      "selectedSectionId": "SERVICE"
    },
    "error": "",
    "yamlSchemaErrorWrapper": {}
  },
  "stepsFactory": { "invocationMap": {}, "stepBank": {}, "stepIconMap": {}, "type": "pipeline-factory" },
  "stagesMap": {
    "Deployment": {
      "name": "Deploy",
      "type": "Deployment",
      "icon": "pipeline-deploy",
      "iconColor": "var(--pipeline-deploy-stage-color)",
      "isApproval": false,
      "openExecutionStrategy": true
    },
    "ci": {
      "name": "Deploy",
      "type": "ci",
      "icon": "pipeline-build",
      "iconColor": "var(--pipeline-build-stage-color)",
      "isApproval": false,
      "openExecutionStrategy": false
    },
    "Pipeline": {
      "name": "Deploy",
      "type": "Pipeline",
      "icon": "pipeline",
      "iconColor": "var(--pipeline-blue-color)",
      "isApproval": false,
      "openExecutionStrategy": false
    },
    "Custom": {
      "name": "Deploy",
      "type": "Custom",
      "icon": "pipeline-custom",
      "iconColor": "var(--pipeline-custom-stage-color)",
      "isApproval": false,
      "openExecutionStrategy": false
    },
    "Approval": {
      "name": "Deploy",
      "type": "Approval",
      "icon": "pipeline-approval",
      "iconColor": "var(--pipeline-approval-stage-color)",
      "isApproval": false,
      "openExecutionStrategy": false
    }
  },
  "allowableTypes": []
}
