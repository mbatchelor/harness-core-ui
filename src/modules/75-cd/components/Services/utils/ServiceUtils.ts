/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { set } from 'lodash-es'
import type { PipelineInfoConfig, ServiceDefinition } from 'services/cd-ng'

export enum ServiceTabs {
  SUMMARY = 'summaryTab',
  Configuration = 'configuration',
  REFERENCED_BY = 'refrencedByTab',
  ActivityLog = 'activityLog'
}

export const DefaultNewStageName = 'Stage Name'
export const DefaultNewStageId = 'stage_id'
export const DefaultNewServiceId = '-1'

export const newServiceState = {
  service: {
    name: '',
    identifier: '',
    description: '',
    gitOpsEnabled: false,
    tags: {},
    serviceDefinition: {
      type: '' as ServiceDefinition['type'],
      spec: {}
    }
  }
}

const DefaultService = {
  serviceDefinition: {
    spec: {}
  },
  gitOpsEnabled: false
}

export const initialServiceState = {
  service: { ...DefaultService }
  // isLoading: false,
  // isBETemplateUpdated: false,
  // isDBInitialized: false,
  // isUpdated: false,
  // isInitialized: false,
  // gitDetails: {},
  // entityValidityDetails: {}
}

export const setNameIDDescription = (draftData: PipelineInfoConfig, updatedData: PipelineInfoConfig): void => {
  set(draftData, 'identifier', updatedData.identifier)
  set(draftData, 'name', updatedData.name)
  set(draftData, 'description', updatedData.description)
  set(draftData, 'tags', updatedData.tags)
}
