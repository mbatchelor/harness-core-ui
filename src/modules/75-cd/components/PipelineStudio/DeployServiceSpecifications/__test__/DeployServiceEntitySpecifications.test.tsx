/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { Formik } from '@harness/uicore'
import { noop } from 'lodash-es'
import { TestWrapper } from '@common/utils/testUtils'
import factory from '@pipeline/components/PipelineSteps/PipelineStepFactory'
import { DeployServiceStep } from '@cd/components/PipelineSteps/DeployServiceStep/DeployServiceStep'
import { KubernetesServiceSpec } from '@cd/components/PipelineSteps/K8sServiceSpec/K8sServiceSpec'
import { ServerlessAwsLambdaServiceSpec } from '@cd/components/PipelineSteps/ServerlessAwsLambdaServiceSpec/ServerlessAwsLambdaServiceSpec'
import {
  PipelineContext,
  PipelineContextInterface
} from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'
import { StageType } from '@pipeline/utils/stageHelpers'
import DeployServiceEntitySpecifications from '../DeployServiceEntitySpecifications'
import { servicesV2Mock } from './servicesMock'
import newServiceEntityPipeline from './overrideSetPipeline.json'

const mockchildren = <div />
const getOverrideContextValue = (): PipelineContextInterface => {
  return {
    ...newServiceEntityPipeline,
    getStageFromPipeline: jest.fn().mockReturnValue({
      stage: {
        stage: {
          name: 'Stage 3',
          identifier: 's3',
          type: StageType.DEPLOY,
          description: '',
          spec: {}
        }
      }
    }),
    updateStage: jest.fn(),
    updatePipeline: jest.fn()
  } as any
}

jest.mock('@common/components/YAMLBuilder/YamlBuilder')
jest.mock('@wings-software/monaco-yaml/lib/esm/languageservice/yamlLanguageService', () => ({
  getLanguageService: jest.fn()
}))

jest.mock('services/cd-ng', () => ({
  useGetServiceList: jest.fn().mockImplementation(() => ({ loading: false, data: servicesV2Mock, refetch: jest.fn() })),
  useGetServiceV2: jest.fn().mockImplementation(() => ({ loading: false, data: {}, refetch: jest.fn() })),
  useGetRuntimeInputsServiceEntity: jest
    .fn()
    .mockImplementation(() => ({ loading: false, data: {}, refetch: jest.fn() }))
}))

// jest.mock('lodash-es', () => ({
//   ...(jest.requireActual('lodash-es') as Record<string, any>),
//   debounce: jest.fn(fn => {
//     fn.cancel = jest.fn()
//     return fn
//   }),
//   noop: jest.fn()
// }))

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
const intersectionObserverMock = () => ({
  observe: () => null,
  unobserve: () => null
})

window.IntersectionObserver = jest.fn().mockImplementation(intersectionObserverMock)

describe('DeployServiceEntitySpecifications', () => {
  beforeAll(() => {
    factory.registerStep(new DeployServiceStep())
    factory.registerStep(new KubernetesServiceSpec())
    factory.registerStep(new ServerlessAwsLambdaServiceSpec())
  })
  test('deployServiceEntitySpecifications renders correctly', () => {
    const { container } = render(
      <TestWrapper>
        <Formik initialValues={{}} onSubmit={noop} formName="deployServiceSpecificationsTest">
          <PipelineContext.Provider value={getOverrideContextValue()}>
            <DeployServiceEntitySpecifications>{mockchildren}</DeployServiceEntitySpecifications>
          </PipelineContext.Provider>
        </Formik>
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })
})
