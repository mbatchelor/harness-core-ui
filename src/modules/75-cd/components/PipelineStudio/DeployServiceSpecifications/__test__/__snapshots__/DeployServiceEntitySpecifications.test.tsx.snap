// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeployServiceEntitySpecifications deployServiceEntitySpecifications renders correctly 1`] = `
<div>
  <div
    class="OverlaySpinner--overlaySpinner"
  >
    <div
      class="deployStage"
    >
      <div
        class="contentSection paddedSection"
      >
        <div
          class="tabHeading"
        >
          cd.pipelineSteps.serviceTab.aboutYourService
        </div>
        <div
          class="bp3-card bp3-elevation-0 Card--card sectionCard"
          id="aboutService"
        >
          <div
            class="OverlaySpinner--overlaySpinner"
          >
            <form
              action="#"
              class="FormikForm--main"
            >
              <div>
                <div
                  class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main formRow StyledProps--flex StyledProps--flex-alignItems-flex-start StyledProps--flex-justifyContent-flex-start"
                >
                  <div
                    class="bp3-form-group"
                    data-id="serviceRef-0"
                  >
                    <label
                      class="bp3-label"
                      for="serviceRef"
                    >
                      <span
                        class="Tooltip--acenter"
                        data-tooltip-id="specifyYourService"
                      >
                        cd.pipelineSteps.serviceTab.specifyYourService
                      </span>
                       
                      <span
                        class="bp3-text-muted"
                      />
                    </label>
                    <div
                      class="bp3-form-content"
                    >
                      <div
                        class="Layout--horizontal StyledProps--main MultiTypeInput--main MultiTypeInput--disabled"
                        style="width: 300px;"
                      >
                        <div
                          class="TextInput--main MultiTypeInput--input"
                        >
                          <div
                            class="bp3-input-group bp3-disabled"
                          >
                            <input
                              class="bp3-input"
                              disabled=""
                              name="serviceRef"
                              placeholder="<+input>"
                              type="text"
                              value="<+input>"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </form>
          </div>
        </div>
        <div
          class="StyledProps--font StyledProps--main StyledProps--margin-top-xxlarge"
        >
          <div />
        </div>
      </div>
    </div>
  </div>
</div>
`;
